#!/usr/bin/env cwl-runner

cwlVersion: v1.0
class: CommandLineTool
label: "pizzly gene fusion detector"
baseCommand: ["pizzly"]
requirements:
    - class: ResourceRequirement
      coresMin: 1
      ramMin: 16000
    - class: DockerRequirement
      dockerPull: "mgibio/pizzly-cwl:0.37.3"
arguments: ["--output", "pizzly"]
inputs:
    kallisto_kmer_size:
        type: int?
        default: 31
        doc: 'k-mer size used when generating the kallisto index'
        inputBinding:
            prefix: "-k"
            position: 1
    transcriptome_annotations:
        type: File
        doc: 'reference transcriptome annotations in GTF format'
        inputBinding:
            prefix: "--gtf"
            position: 2
    reference_transcriptome:
        type: File
        doc: 'reference transcriptome in FASTA format'
        inputBinding:
            prefix: "--fasta"
            position: 3
    fusion_evidence:
        type: File
        doc: 'fusion evidence generated by kallisto'
        inputBinding:
            position: 4
outputs:
    unfiltered_fusion_seqs:
        type: File
        outputBinding:
            glob: "pizzly.unfiltered.fusions.fasta"
    unfiltered_fusions_json:
        type: File
        outputBinding:
            glob: "pizzly.unfiltered.json"
    filtered_fusion_seqs:
        type: File
        outputBinding:
            glob: "pizzly.fusions.fasta"
    filtered_fusions_json:
        type: File
        outputBinding:
            glob: "pizzly.json"
